File: 03_visitor/erlang/subsumed_visitor.erl
#!/usr/bin/env escript

-record (cylinder, {baseRadius = 0, height = 0}).
-record (sphere, {radius = 0}).

volume(#cylinder {baseRadius = BaseRadius, height = Height}) ->
	math:pi() * math:pow(BaseRadius, 2) * Height;
volume(#sphere {radius = Radius}) ->
	4.0 / 3.0 * math:pi() * math:pow(Radius, 3);
volume(List = [_|_]) -> 
  lists:foldl(fun(Shape, Acc) -> Acc + volume(Shape) end, 0, List);
volume([]) -> 0;
volume(_) -> not_ok.

surfaceArea(#cylinder {baseRadius = BaseRadius, height = Height}) ->
  BaseArea = math:pi() * math:pow(BaseRadius, 2),
  BaseCircumference = 2 * math:pi() * BaseRadius,
  2 * BaseArea + BaseCircumference * Height;
surfaceArea(#sphere {radius = Radius}) ->
  4.0 * math:pi() * math:pow(Radius, 2);
surfaceArea(List = [_|_]) -> 
  lists:foldl(fun(Shape, Acc) -> Acc + surfaceArea(Shape) end, 0, List);
surfaceArea([]) -> 0;
surfaceArea(_) -> not_ok.

render(openGL, #cylinder {}) ->
  io:format("~p~n", ["OpenGL: rendering cylinder"]);
render(openGL, #sphere {}) ->
  io:format("~p~n", ["OpenGL: rendering sphere"]);
render(svg, #cylinder {}) ->
  io:format("~p~n", ["SVG: rendering cylinder"]);
render(svg, #sphere {}) ->
  io:format("~p~n", ["SVG: rendering sphere"]);
render(Platform, List = [_|_]) -> 
  io:format("Rendering Composite...~n"),
  lists:foreach(fun(Shape) -> render(Platform, Shape) end, List);
render(_, _) -> not_ok.

main([]) ->
  Cylinder = #cylinder {baseRadius = 10, height = 10},
  Sphere = #sphere {radius = 10},
  Composite = [Cylinder, Sphere],
  io:format("Volume of Cylinder ~p~n", [volume(Cylinder)]),
  io:format("Volume of Sphere ~p~n", [volume(Sphere)]),
  io:format("Volume of Composite ~p~n", [volume(Composite)]),
  % io:format("Volume of 2 ~p~n", [volume(2)]),
  io:format("Surface Area of Cylinder ~p~n", [surfaceArea(Cylinder)]),
  io:format("Surface Area of Sphere ~p~n", [surfaceArea(Sphere)]),
  io:format("Surface Area of Composite ~p~n", [surfaceArea(Composite)]),
  % io:format("Surface Area of 2 ~p~n", [surfaceArea(2)]),
  io:format("~p~n", [render(openGL, Composite)]),
  io:format("~p~n", [render(svg, Composite)]),
  io:format("DONE").

File: 03_visitor/clojure/subsumed_visitor.clj
(defrecord Cylinder [baseRadius height])
(defrecord Sphere [radius])

(defmulti volume class)
(defmethod volume Cylinder [{:keys [baseRadius height]}]
  (* (. Math PI) (Math/pow baseRadius 2) height))
(defmethod volume Sphere [{:keys [radius]}] 
  (* (double (/ 4 3)) (. Math PI) (Math/pow radius 3)))
(defmethod volume clojure.lang.IPersistentVector [shapes]
  (reduce + 0 (map volume shapes)))
(defmethod volume :default [s] nil)

(defmulti surface-area class)
(defmethod surface-area Cylinder [{:keys [baseRadius height]}]
	(let [base-area (* (. Math PI) (Math/pow baseRadius 2))
	    base-circumference (* 2 (. Math PI) baseRadius)]
	    (+ (* 2 base-area) (* base-circumference height))))
(defmethod surface-area Sphere [{:keys [radius]}] 
  (* 4 (. Math PI) (Math/pow radius 2)))
(defmethod surface-area clojure.lang.IPersistentVector [shapes]
  (reduce + 0 (map surface-area shapes)))
(defmethod surface-area :default [s] nil)

(def cylinder (Cylinder. 10 10))
(def sphere (Sphere. 10))
(def composite [cylinder sphere])

(println (volume cylinder)) ; 3141.59
(println (volume sphere)) ;4188.79
(println (volume composite)) 
(println (volume 2)) ;nil

(println (surface-area cylinder)) ;1256.63
(println (surface-area sphere)) ;1256.63
(println (surface-area composite)) 
(println (surface-area 2)) ;nil

(defrecord OpenGL [])
(defrecord SVG [])

(defmulti render 
  (fn [platform shape] [(class platform) (class shape)]))
  
(defmethod render [OpenGL Cylinder] [platform shape] (println "OpenGL: Rendering Cylinder"))
(defmethod render [OpenGL Sphere]   [platform shape] (println "OpenGL: Rendering Sphere"))
(defmethod render [SVG Cylinder] [platform shape] (println "SVG: Rendering Cylinder"))
(defmethod render [SVG Sphere]   [platform shape] (println "SVG: Rendering Sphere"))
(defmethod render [Object clojure.lang.IPersistentVector] [platform shapes] 
  (println "Rendering Composite...")
  (doall (map #(render platform %) shapes)))
(defmethod render :default [platform shape] nil)

(render (OpenGL.) composite)
(render (SVG.) composite)
(println "DONE")
